### DISCUSSION TITLE
Returning multiple values using WasmEdge C SDK

### DISCUSSION THREAD

**Discussion by mtobin-tdab** [WasmEdge/WasmEdge#3344] [General]
Bear in mind I am new to c++ and am trying to get my head around how the WasmEdge C SDK works.

I am executing a wasm function using `WasmEdge_VMExecute` - the function takes `ReturnLen` implying that we can return multiple values although I am not sure how this would look on the rust side since only integers can be returned.

Could you please advise on this? Below is my simple rust function and SDK code:

```rust
#[no_mangle]
pub extern "C" fn return_pointer_length() -> Vec<usize> {
    vec![1, 2]
}
```

```cpp
WasmEdge_Value funcRet[2];
Res = WasmEdge_VMExecute(VMCxt, FuncName, NULL, 0, funcRet, 2);
outputPointer = (uint32_t) WasmEdge_ValueGetI32(funcRet[0]);
uint32_t length = (uint32_t) WasmEdge_ValueGetI32(funcRet[1]);
```

This obviously doesn't work but hopefully illustrates what I am attempting.

**Comment by hydai** [comment]
Please check the rust compiler and the output to learn the memory layout. All the hacks use the same idea: read the binary generated by the `rustc` and try to figure out how to use the golang/rust to decode/encode its memory section.
Empower yourself by using the `wasm2wat` tool to convert the compiled wasm file into the text format. This will allow you to check the function signature and then experiment with the parameters to understand their significance. This hands-on approach will give you the confidence to interact with the C SDK effectively.
